### DISCUSSION TITLE
Enable/disable metrics in a custom extension

### DISCUSSION THREAD

**Discussion by Sirz3chs** [quarkusio/quarkus#31176] [Q&A] [answered]
I'm actually writing an extension to collect custom logs from endpoints and send them elsewhere with `quarkus-rest-client-reactive`. I wanted to add the possibility to enable/disable metrics on rest client but i can't figure out how exactly manage the build time.

I just have a simple build time config:
```
@ConfigRoot(name = "custom-logs", phase = ConfigPhase.BUILD_TIME)
public class CustomLogsBuildTimeConfig {

    @Getter
    @ConfigItem(name = "metrics.enabled", defaultValue = "true")
    protected boolean metricsEnabled;
}
```
I've started to write this in my processor:
```
@BuildStep
@Record(ExecutionTime.STATIC_INIT)
void metrics(final CustomLogsBuildTimeConfig config, final Optional<MetricsCapabilityBuildItem> metrics) {
    if (config.metricsEnabled() && metrics.isPresent()) {
        ...
    }
}
```

I suppose, according to the [quarkus documentation](https://quarkus.io/guides/writing-extensions#extension-metrics), that i'm in the case:
`An underlying library used by the extension is using a specific Metrics API directly (either MP Metrics, Micrometer, or some other).`

How am i suppose to configure metrics only for my extension ?

[a comment by an automated account is not shown]

**Comment by gsmet** [comment]
@Sgitario is there a way to enable/disable metrics for the REST Clients (legacy or reactive)? I couldn't find an associated configuration property.

Not sure if it would be a runtime property though and if it would be possible to use `RunTimeConfigurationDefaultBuildItem`.

**Reply by Sgitario** [reply]
As far as I can tell, REST Client does not generate metrics that we could enable/disable either in REST Clients (legacy or reactive).

**Reply by ebullient** [reply]
Yes, there is a client configuration for metrics (reactive and not): [quarkus.micrometer.binder.http-client.enabled](https://quarkus.io/guides/micrometer#quarkus-micrometer_quarkus.micrometer.binder.http-client.enabled)

```
quarkus.micrometer.binder.http-client.enabled=false
```

**Comment by ebullient** [comment, marked as answer]
For the original poster: enabling/disabling client metrics is a little more complicated for http specifically. Metrics are already collected for REST Client (with micrometer) for Resteasy and Resteasy Reactive... if you're working against the micrometer extension, you might be better off following a pattern like this: 

https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/instrument/logging/LoggingMeterRegistry.java  

Your extension can then focus on enabling or disabling the construction of a custom MeterRegistry bean. 

As far as testing for enabled metrics at build time, you'd want to look at a combination of config, you would want to do something akin to this: 

https://github.com/quarkusio/quarkus/blob/e4b4aeaf6989b83b2e2ac281020c50585cdeee32/extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/binder/HttpBinderProcessor.java#L50

Here is an example processor that splits to do different things if micrometer is present: 
https://github.com/quarkusio/quarkus/blob/e4b4aeaf6989b83b2e2ac281020c50585cdeee32/extensions/jaeger/deployment/src/main/java/io/quarkus/jaeger/deployment/JaegerProcessor.java#L38

**Reply by Sirz3chs** [reply]
Ok, thanks for the explanation, i'll try to implement this in my project
